src/backend/distributed/operations/repair_shards.c

```diff
@@ -131,19 +148,155 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
+ * short duration almost only when the metadata is actually being updated. This option
+ * is currently only available in Citus Enterprise.
+ *
+ * After successful move operation, shards in the source node gets deleted. If the move
+ * fails at any point, this function throws an error, leaving the cluster without doing
+ * any changes in source node or target node.
  */
 Datum
 master_move_shard_placement(PG_FUNCTION_ARGS)
 {
+	int64 shardId = PG_GETARG_INT64(0);
+	char *sourceNodeName = text_to_cstring(PG_GETARG_TEXT_P(1));
+	int32 sourceNodePort = PG_GETARG_INT32(2);
+	char *targetNodeName = text_to_cstring(PG_GETARG_TEXT_P(3));
+	int32 targetNodePort = PG_GETARG_INT32(4);
+	Oid shardReplicationModeOid = PG_GETARG_OID(5);
+
+
+	ListCell *colocatedTableCell = NULL;
+	ListCell *colocatedShardCell = NULL;
+
+
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid relationId = RelationIdForShard(shardId);
+	ErrorIfMoveCitusLocalTable(relationId);
+
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	Oid distributedTableId = shardInterval->relationId;
+
+	List *colocatedTableList = ColocatedTableList(distributedTableId);
+	List *colocatedShardList = ColocatedShardIntervalList(shardInterval);
+
+	foreach(colocatedTableCell, colocatedTableList)
+	{
+		Oid colocatedTableId = lfirst_oid(colocatedTableCell);
+		char relationKind = '\0';
+
+		/* check that user has owner rights in all co-located tables */
+		EnsureTableOwner(colocatedTableId);
+
+		/*
+		 * Block concurrent DDL / TRUNCATE commands on the relation. Similarly,
+		 * block concurrent master_move_shard_placement() on any shard of
+		 * the same relation. This is OK for now since we're executing shard
+		 * moves sequentially anyway.
+		 */
+		LockRelationOid(colocatedTableId, ShareUpdateExclusiveLock);
+
+		relationKind = get_rel_relkind(colocatedTableId);
+		if (relationKind == RELKIND_FOREIGN_TABLE)
+		{
+			char *relationName = get_rel_name(colocatedTableId);
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot repair shard"),
+							errdetail("Table %s is a foreign table. Repairing "
+									  "shards backed by foreign tables is "
+									  "not supported.", relationName)));
+		}
+	}
+
+	/* we sort colocatedShardList so that lock operations will not cause any deadlocks */
+	colocatedShardList = SortList(colocatedShardList, CompareShardIntervalsById);
+	foreach(colocatedShardCell, colocatedShardList)
+	{
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		uint64 colocatedShardId = colocatedShard->shardId;
+
+		EnsureShardCanBeCopied(colocatedShardId, sourceNodeName, sourceNodePort,
+							   targetNodeName, targetNodePort);
+	}
+
+	char shardReplicationMode = LookupShardTransferMode(shardReplicationModeOid);
+	if (shardReplicationMode == TRANSFER_MODE_FORCE_LOGICAL)
+	{
+		ereport(WARNING, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
```

Comment:
I thought we agreed to error here and later introduce a guc if this becomes a problem

Reply:
yes, fixed now, and added tests